### PR TITLE
[TP] Add prepareInput and output for input/output DTensor layout annotation in the parent module in TP API

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111176
* __->__ #111166
* #111160

In some use cases, we found that users might want to annote the input/output DTensor layout for the parent module rather than the submodule whose parameters are to be distributed so that we want to have these two class for users to annote input/output DTensor layouts so that we register pre-FWD/FWD hook for the TP-lized module.

### CODE REVIEW

File: torch/distributed/tensor/parallel/style.py [hunk 1]
```diff
@@ -565,3 +569,40 @@ def __init__(
             if _prepare_output is not None
             else _get_prepare_output(output_layouts, use_local),
         )
+
+
+class PrepareModuleInput(ParallelStyle):
+    """
+    Only used to specify the input distribute spec for a module.
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:576]: we should make this documentation clearer, we should mention sth like:
PrepareModuleInput allows user to annotate the input torch.Tensor layouts to create DTensors base on input_layouts, and prepare the input DTensors be transformed to a desired layout base on output layouts. 

We should also explain each kwarg separately, and possibly add a small example demonstrate: annotate input torch.Tensor as a Sharded DTensor, then transform it to a Replicated DTensor

Comment (fduwjj) [reply torch/distributed/tensor/parallel/style.py]: Added comments, args doc and examples.

File: torch/distributed/tensor/parallel/style.py [hunk 2]
```diff
@@ -589,3 +591,102 @@ def __init__(
...
+        >>>     ...,
+        >>>     parallelize_plan=parallelize_plan,
+        >>> )
+        >>> ...
+        """
+        super().__init__(
+            input_layouts=input_layouts,
+            output_layouts=output_layouts,
+            use_local=use_local,
+            _prepare_input=_get_prepare_input(
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:645]: a bit surprised it works as `_prepare_input/output` are positional args and it should appear in the front..

File: torch/distributed/tensor/parallel/style.py [hunk 3]
```diff
@@ -589,3 +591,102 @@ def __init__(
...
+
+        Returns:
+            None.
+
+        Example::
+        >>> # xdoctest: +SKIP(failing)
+        >>> from torch.distributed.tensor.parallel import parallelize_module, PrepareModuleInput
+        >>> ...
+        >>> parallelize_plan = {
+        >>>     "attn": PrepareModuleInput(),   # The input of attn will be converted to Sharded DTensor.
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:631]: nit: input of attn will be converted to Sharded DTensor and redistributed to Replicated DTensor?

File: torch/distributed/tensor/parallel/style.py [hunk 4]
```diff
@@ -589,3 +591,102 @@ def __init__(
...
+
+    When the output is not a :class:`DTensor`, if no layout is specified, it will be
+    a no-op. Otherwise, it will throw an error.
+
+    Example::
+    >>> # xdoctest: +SKIP(failing)
+    >>> from torch.distributed.tensor.parallel import parallelize_module, PrepareModuleOutput
+    >>> ...
+    >>> parallelize_plan = {
+    >>>     "submodule": PrepareModuleOutput(),   # The input of attn will be converted to Sharded tensor.
```
Comment (wanchaol) [inline torch/distributed/tensor/parallel/style.py:669]: nit: the output of attn will be converted to Replicated DTensor if it's not a DTensor, then redistributed to Sharded local tensor

**Review by wanchaol — APPROVED** [review]
lgtm, small nits inlined

File: torch/distributed/tensor/parallel/style.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/tensor/parallel/style.py:673]: but positional args can also be passed in as kwargs I guess?

File: docs/source/distributed.tensor.parallel.rst [hunk 5]
```diff
@@ -53,6 +53,9 @@ used for input/output preparation:
 .. autofunction::  make_output_shard_1d
 .. autofunction::  make_output_tensor
 
+.. autofunction::  PrepareModuleInput
```
Comment (fduwjj) [inline docs/source/distributed.tensor.parallel.rst:56]: Add it here to pass CI first, Will have more doc changes in the next PR https://github.com/pytorch/pytorch/pull/111176

**Pull request closed without merging**